CMakeLists.txt

```diff
@@ -109,16 +109,23 @@ else()
 
   # No config file for this
   if(WITH_GFLAGS)
-    find_package(gflags REQUIRED)
+    option(GFLAGS_USE_TARGET_NAMESPACE "Use gflags import target with namespace." ON)
+    find_package(gflags CONFIG)
+    if(NOT gflags_FOUND)
+      find_package(gflags REQUIRED)
```

Comment:
could you remind me what we might run into if we just call
```cmake
find_package(gflags REQUIRED)
```
without doing the "CONFIG" then "MODULE" dance?

Reply:
Because the FindXXX we provided is for fallback when XXXConfig is missing, not for overwriting the behavior of XXXConfig.
I think CMake chooses MODULE > CONFIG for backward compatibility (CMake 2 code expecting only FindXXX before XXXConfig is invented still gets FindXXX in CMake 3).

Another option is to set `CMAKE_FIND_PACKAGE_PREFER_CONFIG`, which changes the default to CONFIG > MODULE implicitly.
I'm not doing that because it also changes the behavior of both caller (RocksDB) and callee (gflags/snappy/... and their deps).

Reply:
Hmmm...Double checked the doc, CONFIG keyword was available much earlier than I expected, but I never seen people use it at that time.

Reply:
> Because the FindXXX we provided is for fallback when XXXConfig is missing, not for overwriting the behavior of XXXConfig.

i agree with you on this. but does it hurt to allow cmake to use the homebrew `Findfoo.cmake` even if `fooConfig` exists ?

Reply:
Yes because usually upstream knows better about their own lib.
For example:
1. foo 1.0 and foo 1.1 may use different dir layout and FindXXX designed for 1.1 won't work for 1.0.
2. Here's a gflags compiled for shared (dll) only, but with both threading/serial version.
If user wants to use `gflags_nothreads.lib`, with upstream CONFIG all it take is to `-DGFLAGS_NOTHREADS=ON`.
The `Findgflags.cmake` in RocksDB repo doesn't work for that.
<img width="839" alt="image" src="https://user-images.githubusercontent.com/5203025/81007933-a5666580-8e84-11ea-9f5f-944cac804e43.png">
3. It's also possible for distro packager to compose special dir layout and export with different set of configs.
Use case include having libs built with multiple compiler runtime (e.g. gcc/llvm/icc, very common in HPC environment).

I think config file is an important feature of cmake ecosystem.
If upstream can tell you what to use, why hard-code it?

Reply:
> 2. Here's a gflags compiled for shared (dll) only, but with both threading/serial version.
>    If user wants to use `gflags_nothreads.lib`, with upstream CONFIG all it take is to `-DGFLAGS_NOTHREADS=ON`.
>    The `Findgflags.cmake` in RocksDB repo doesn't work for that.

very true, we are not able to compete with upstream config in this perspective.